src/core/Akka.Remote/Transport/DotNetty/BatchWriter.cs

```diff
@@ -6,181 +6,242 @@
+                return true;
+            } // else we'll flush when the read completes
+
+            // didn't flush
+            return false;
         }
 
-        /// <summary>
-        /// Toggle for turning this feature on or off.
-        /// </summary>
-        /// <remarks>
-        /// Defaults to <c>true</c>.
-        /// </remarks>
-        public bool EnableBatching { get; }
+        // cache the delegate
+        private static readonly Func<object, bool> FlushOp = obj => TryFlush((FlushConsolidationHandler)obj);
 
-        /// <summary>
-        /// The maximum amount of buffered writes that can be buffered before flushing I/O.
-        /// </summary>
-        /// <remarks>
-        /// Defaults to 30.
-        /// </remarks>
-        public int MaxPendingWrites { get; }
+        public FlushConsolidationHandler() : this(DefaultExplicitFlushAfterFlushes, true){}
 
-        /// <summary>
-        /// In the event of low-traffic channels, the maximum amount of time we'll wait before flushing writes.
-        /// </summary>
-        /// <remarks>
-        /// Defaults to 40 milliseconds.
-        /// </remarks>
-        public TimeSpan FlushInterval { get; }
+        public FlushConsolidationHandler(int explicitFlushAfterFlushes) : this(explicitFlushAfterFlushes, true) { }
 
-        /// <summary>
-        /// The maximum number of outstanding bytes that can be written prior to a flush.
-        /// </summary>
-        /// <remarks>
-        /// Defaults to 16kb.
-        /// </remarks>
-        public long MaxPendingBytes { get; }
-    }
+        public FlushConsolidationHandler(int explicitFlushAfterFlushes, bool consolidateWhenNoReadInProgress)
+        {
+            ExplicitFlushAfterFlushes = explicitFlushAfterFlushes;
+            ConsolidateWhenNoReadInProgress = consolidateWhenNoReadInProgress;
+        }
 
-    /// <summary>
-    /// INTERNAL API.
-    /// 
-    /// Responsible for batching socket writes together into fewer sys calls to the socket.
-    /// </summary>
-    internal class BatchWriter : ChannelHandlerAdapter
-    {
-        private class FlushCmd
+        public override void HandlerAdded(IChannelHandlerContext context)
         {
-            public static readonly FlushCmd Instance = new FlushCmd();
-            private FlushCmd() { }
+            _context = context;
         }
 
-        public readonly BatchWriterSettings Settings;
-        public readonly IScheduler Scheduler;
-        private ICancelable _flushSchedule;
+        public override void Flush(IChannelHandlerContext context)
+        {
+            if (_readInProgress)
+            {
+                // If there is still a read in progress we are sure we will see a channelReadComplete(...) call. Thus
+                // we only need to flush if we reach the explicitFlushAfterFlushes limit.
+                if (++_flushPendingCount == ExplicitFlushAfterFlushes)
+                {
+                    FlushNow(context);
```

Comment:
So when reading is complete we will flush in `ChannelReadComplete` anyway. But we are trying to flush right here if there are too many flushes are pending? Are we able to flush while reading is in process? Or this is safe?

Reply:
So if a read is sitting in the channel pipeline right now we assume it means that some data is being read by the current application and that might, presumably, be used to produce a response right away - we're trying to batch those writes together into as few flushes as possible and when the read "completes" (we've finished reading all of the data currently inside the buffer) it's safe to flush any writes that are currently pending. This helps reduce latency in lower traffic system AND increases throughput in higher traffic systems.